control-plane/agents/src/bin/core/controller/reconciler/volume/garbage_collector.rs

```diff
@@ -345,62 +338,42 @@ async fn disown_non_reservable_replicas(
-                            && !replica
-                                .as_ref()
-                                .owners
-                                .nexuses()
-                                .iter()
-                                .any(|p| p != nexus.uuid())
-                            && !is_replica_healthy(
-                                context,
-                                &mut nexus_info,
-                                replica.as_ref(),
-                                volume.as_ref(),
-                            )
-                            .await?
-                        {
-                            volume.warn_span(|| tracing::warn!(replica.uuid = %replica.as_ref().uuid, "Attempting to disown non reservable replica"));
-
-                            // Since its a local replica of a shutting down nexus i.e. a failed
-                            // shutdown nexus we don't want it
-                            // to be a part of anything. Once this replica has no owner, the
-                            // garbage collector will remove the nexus. Even if the clean up is
-                            // delayed we since this replica is
-                            // anyhow not a part of the volume, we should be safe.
-                            match replica
-                                .remove_owners(
-                                    context.registry(),
-                                    &ReplicaOwners::new_disown_all(),
-                                    true,
-                                )
-                                .await
-                            {
-                                Ok(_) => {
-                                    volume.info_span(|| tracing::info!(replica.uuid = %replica.as_ref().uuid, "Successfully disowned non reservable replica"));
-                                }
-                                Err(error) => {
-                                    volume.error_span(|| tracing::error!(replica.uuid = %replica.as_ref().uuid, "Failed to disown non reservable replica"));
-                                    results.push(Err(error));
-                                }
-                            }
-                        }
-                    }
+            let Some(replica_uri) = child.as_replica() else {
+                continue;
+            };
+            if !replica_uri.uri().is_local() {
+                continue;
+            }
+            let Ok(replica) = context.specs().replica(replica_uri.uuid()).await else {
+                continue;
+            };
+            if !replica.as_ref().dirty()
+                && replica.as_ref().owners.owned_by(volume.uuid())
+                && !replica
+                    .as_ref()
+                    .owners
+                    .nexuses()
+                    .iter()
+                    .any(|p| p != nexus.uuid())
+                && !is_replica_healthy(context, &mut nexus_info, replica.as_ref(), volume.as_ref())
```

Comment:
I wonder do we really need to have health check of replica here? Irrespective of it's healthy or not, we can probably disown it from volume but not nexus, because eventually it is bound to be destroyed when nexus shutdown/cleanup happens.

Reply:
I guess it's just a safety gate, because is shouldn't really be healthy